framework/src/transfers/MultiAppMeshFunctionTransfer.C

```diff
@@ -63,289 +62,347 @@ MultiAppMeshFunctionTransfer::execute()
+          if (point_index_map[i_proc].find(key) == point_index_map[i_proc].end())
+            continue;
+          unsigned int i_pt = point_index_map[i_proc][key];
+
+          // Ignore this proc if it's app has a higher rank than the
+          // previously found lowest app rank.
+          if (_direction == FROM_MULTIAPP)
           {
-            Node * node = *node_it;
-
-            if (node->n_dofs(to_sys_num, to_var_num) > 0) // If this variable has dofs at this node
-            {
-              // See if this node falls in this bounding box
-              if (app_box.contains_point(*node-app_position))
-              {
-                // The zero only works for LAGRANGE!
-                dof_id_type dof = node->dof_number(to_sys_num, to_var_num, 0);
-
-                MPI_Comm swapped = Moose::swapLibMeshComm(_multi_app->comm());
-                Real from_value = from_func(*node-app_position);
-                Moose::swapLibMeshComm(swapped);
-
-                if (from_value != OutOfMeshValue)
-                  to_solution->set(dof, from_value);
-                else if (_error_on_miss)
-                  mooseError("Point not found! " << *node-app_position <<std::endl);
-              }
-            }
+            if (incoming_app_ids[i_proc][i_pt] >= lowest_app_rank)
+              continue;
           }
-        }
-        else // Elemental
-        {
-          MeshBase::const_element_iterator elem_it = to_mesh.elements_begin();
-          MeshBase::const_element_iterator elem_end = to_mesh.elements_end();
 
-          for (; elem_it != elem_end; ++elem_it)
-          {
-            Elem * elem = *elem_it;
+          // Ignore this proc if the point was actually outside its meshes.
+          if (incoming_evals[i_proc][i_pt] == OutOfMeshValue)
+            continue;
 
-            if (elem->n_dofs(to_sys_num, to_var_num) > 0) // If this variable has dofs at this elem
-            {
-              Point centroid = elem->centroid();
-
-              // See if this elem falls in this bounding box
-              if (app_box.contains_point(centroid-app_position))
-              {
-                // The zero only works for LAGRANGE!
-                dof_id_type dof = elem->dof_number(to_sys_num, to_var_num, 0);
-
-                MPI_Comm swapped = Moose::swapLibMeshComm(_multi_app->comm());
-                Real from_value = from_func(centroid-app_position);
-                Moose::swapLibMeshComm(swapped);
-
-                if (from_value != OutOfMeshValue)
-                  to_solution->set(dof, from_value);
-                else if (_error_on_miss)
-                  mooseError("Point not found! " << centroid-app_position << std::endl);
-              }
-            }
-          }
+          best_val = incoming_evals[i_proc][i_pt];
+          point_found = true;
         }
-        delete serialized_from_solution;
-      }
 
-      to_solution->close();
-      to_sys.update();
+        if (_error_on_miss && ~ point_found)
```

Comment:
`~` is bitwise NOT, which probably does work for bool, but it is more standard to use logical negation, `!`